dace/transformation/dataflow/hbm_transform.py

```diff
@@ -0,0 +1,198 @@
+            other_node = path[0].src
+        elif edge_index == len(path) - 1:
+            is_write = False
+            other_node = path[-1].dst
+        else:
+            raise ValueError("The provided edge is not the innermost")
+
+        if isinstance(other_node, nd.AccessNode):
+            fwtasklet = state.add_tasklet("fwtasklet", set(["_in"]),
+                                          set(["_out"]), "_out = _in")
+            state.remove_edge_and_connectors(inner_edge)
+            target_other_subset = mem.other_subset
+            mem.other_subset = None
+            if is_write:
+                inner_edge = state.add_edge(fwtasklet, '_out', inner_edge.dst,
+                                            inner_edge.dst_conn, mem)
+                state.add_edge(
+                    other_node, path[0].src_conn, fwtasklet, "_in",
+                    memlet.Memlet(other_node.data, subset=target_other_subset))
+            else:
+                inner_edge = state.add_edge(inner_edge.src, inner_edge.src_conn,
+                                            fwtasklet, '_in', mem)
+                state.add_edge(
+                    fwtasklet, "_out", other_node, path[-1].dst_conn,
+                    memlet.Memlet(other_node.data, subset=target_other_subset))
+
+        utils.update_path_subsets(state, inner_edge, new_subset)
+
+    def _update_array_hbm(self, array_name: str, new_memory: str, new_bank: str,
+                          sdfg: SDFG):
+        """
+        Updates bank assignments for the array
+        """
+        desc = sdfg.arrays[array_name]
+        old_memory = None
+        if 'memorytype' in desc.location and desc.location[
+                "memorytype"] is not None:
+            old_memory = desc.location["memorytype"]
+        if new_memory == "HBM":
+            low, high = fpga.get_multibank_ranges_from_subset(new_bank, sdfg)
+        else:
+            low, high = int(new_bank), int(new_bank) + 1
+        if (old_memory is None or old_memory == "DDR") and new_memory == "HBM":
+            desc = utils.update_array_shape(sdfg, array_name,
+                                            (high - low, *desc.shape))
+        elif old_memory == "HBM" and (new_memory == "DDR"
+                                      or new_memory is None):
+            desc = utils.update_array_shape(array_name, *(list(desc.shape)[1:]))
+        elif old_memory == "HBM" and new_memory == "HBM":
+            new_shape = list(desc.shape)
+            new_shape[0] = high - low
+            desc = utils.update_array_shape(sdfg, array_name, new_shape)
+        desc.location["memorytype"] = new_memory
+        desc.location['bank'] = new_bank
+        desc.storage = dtypes.StorageType.FPGA_Global
+
+    @staticmethod
+    def can_be_applied(self, graph: Union[SDFG, SDFGState],
+                       candidate: Dict['PatternNode', int], expr_index: int,
+                       sdfg: SDFG, strict: bool) -> bool:
+        return True
```

Comment:
No way this is always true? For example, you seem to expect that each accessed data node only has one memlet connected to it. What if you already applied this transformation? Then it also shouldn't apply.

Reply:
With the current implementation, it seems like it also shouldn't match if no arrays have been specified